python/cog/server/worker.py

```diff
@@ -292,41 +292,38 @@ def run(self) -> None:
         # Initially, we ignore SIGUSR1.
         signal.signal(signal.SIGUSR1, signal.SIG_IGN)
 
-        redirector = StreamRedirector(
+        async_redirector = AsyncStreamRedirector(
             callback=self._stream_write_hook,
             tee=self._tee_output,
         )
 
-        # TODO: support async setup? see where `redirector` is redefined below if the predict is async
-        with redirector:
-            self._setup(redirector)
+        with async_redirector:
```

Comment:
Do we not need this to be conditional on the same conditions as predict?

Reply:
It _was_, but this is intentionally different now since the async redirector will work for sync and async predict and _does not_ attempt to do low-level things with file descriptors. This means that setup will not capture writes from outside Python, which is judged (by me and @nickstenning) to be an acceptable trade-off.

Reply:
Ah. I see the problem. We might not have a predictor until after setup, so we're not able to work out if we need async or not before we run it.

Reply:
Yep that!